### PR TITLE
Add Course::Registration model

### PR DESCRIPTION
**Pull request by DIdiHL**
Fill in course user's tests and create enrol request's specs.

### CODE REVIEW

File: db/migrate/20150316144315_remove_name_from_course_users.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class RemoveNameFromCourseUsers < ActiveRecord::Migration
+  def change
+    remove_column :course_users, :name
```
Comment (allenwq) [inline db/migrate/20150316144315_remove_name_from_course_users.rb]: I think this field might be essential, because we want do course specific name in v2, Thus teachers can prevent student from change their names in course. I've a email about this, will forward you.

File: app/models/course/enrol_request.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+                                  CourseUser.roles[:owner]]) }
+
+  enum role: CourseUser.roles
+
+  belongs_to :course
+  belongs_to :user
+
+  def approve!
+    unless CourseUser.where(course_id: course_id, user_id: user_id).empty?
+      return
```
Comment (allenwq) [inline app/models/course/enrol_request.rb]: return should be put on the previous line <code>return unless xxx</code>, otherwise I think rubocop will complain this.

File: db/migrate/20150316135303_create_course_enrol_requests.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+class CreateCourseEnrolRequests < ActiveRecord::Migration
+  def change
+    create_table :course_enrol_requests do |t|
+      t.belongs_to :user
+      t.belongs_to :course
+      t.integer :role
+      t.index [:user_id, :course_id], unique: true
+
+      t.datetime :deleted_at
```
Comment (allenwq) [inline db/migrate/20150316135303_create_course_enrol_requests.rb]: <code>t.datetime :deleted_at, index: true</code>

File: db/migrate/20150316135303_create_course_enrol_requests.rb [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+      t.index :deleted_at
+      t.userstamps null: false, foreign_key: { references: :users }
+      t.timestamps
```
Comment (allenwq) [inline db/migrate/20150316135303_create_course_enrol_requests.rb]: we also use <code>null: false </code> for timestamps.

File: db/migrate/20150316135303_create_course_enrol_requests.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
+class CreateCourseEnrolRequests < ActiveRecord::Migration
+  def change
+    create_table :course_enrol_requests do |t|
+      t.belongs_to :user
```
Comment (allenwq) [inline db/migrate/20150316135303_create_course_enrol_requests.rb]: should put a <code>null: false </code> for all the columns which aren't supposed to be null.

File: spec/models/course/enrol_request_spec.rb [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+      subject { create(:course_enrol_request, course: course, user: user, role: :student) }
+      before do
+        create(:course_enrol_request, course: course, user: user, role: :student)
+      end
+      it 'fails' do
+        expect { subject }.to change(Course::EnrolRequest, :count).by(0).and raise_error
+      end
+    end
+
+    describe 'get student scope' do
```
Comment (allenwq) [inline spec/models/course/enrol_request_spec.rb]: Scope is actually a class methods, think we should use some thing like <code>describe '.student'</code> ? @lowjoel

File: Gemfile [hunk 7]
```diff
@@ -44,6 +44,8 @@ gem 'slim-rails'
 gem 'high_voltage'
 # Paginator for Rails
 gem 'kaminari'
+# Soft deletion
+gem 'paper_trail', '~> 4.0.0.beta'
```
Comment (lowjoel) [inline Gemfile]: Configure paper_trail in a separate PR.

File: app/controllers/course/courses_controller.rb [hunk 8]
```diff
@@ -8,7 +8,7 @@ def new #:nodoc:
   end
 
   def create #:nodoc:
-    @course.course_users.build(user: current_user, name: 'name', role: :owner)
+    @course.course_users.build(user: current_user, role: :owner)
```
Comment (lowjoel) [inline app/controllers/course/courses_controller.rb]: don't we want the student to have an editable name?

File: app/models/course.rb [hunk 9]
```diff
@@ -9,6 +9,7 @@ class Course < ActiveRecord::Base
   belongs_to :creator, class_name: User.name
   has_many :course_users, inverse_of: :course, dependent: :destroy
   has_many :users, through: :course_users
+  has_many :enrol_requests
```
Comment (lowjoel) [inline app/models/course.rb]: does the inverse get properly detected?

File: app/models/course/enrol_request.rb [hunk 10]
```diff
@@ -0,0 +1,25 @@
+class Course::EnrolRequest < ActiveRecord::Base
+  stampable
+  has_paper_trail on: [:destroy]
```
Comment (lowjoel) [inline app/models/course/enrol_request.rb]: I'm not sure all these kinds of data needs to have an explicit trail.. hm.

What do we want to capture edits/modifications of, and what do we not want to worry about?

File: app/models/course/enrol_request.rb [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+  validates :course_id, uniqueness: { scope: :user_id }
```
Comment (lowjoel) [inline app/models/course/enrol_request.rb]: i think this is the other way around. but if we already have a unique constraint on course_id -> user_id, then this is not needed because the validations gem will propagate it for us automatically.

File: app/models/course/enrol_request.rb [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+
+  scope :student, -> { where(role: CourseUser.roles[:student]) }
```
Comment (lowjoel) [inline app/models/course/enrol_request.rb]: scopes should be plural.

File: app/models/course/enrol_request.rb (hunk 2 above)
Comment (lowjoel) [inline app/models/course/enrol_request.rb]: this function is racy. approve can still error when more than one user approves the request at the same time.

File: app/models/course_user.rb [hunk 13]
```diff
@@ -2,4 +2,9 @@ class CourseUser < ActiveRecord::Base
   belongs_to :user, inverse_of: :course_users
   belongs_to :course, inverse_of: :course_users
   enum role: { student: 0, teaching_assistant: 1, manager: 2, owner: 3 }
+  scope :student, -> { where(role: roles[:student]) }
```
Comment (lowjoel) [inline app/models/course_user.rb]: scopes should be plural.

File: app/models/user.rb [hunk 14]
```diff
@@ -9,6 +9,7 @@ class User < ActiveRecord::Base
   has_many :instances, through: :instance_users
   has_many :course_users, inverse_of: :user, dependent: :destroy
   has_many :courses, through: :course_users
+  has_many :enrol_request
```
Comment (lowjoel) [inline app/models/user.rb]: plural. inverse detected?

File: db/migrate/20150316135303_create_course_enrol_requests.rb [hunk 15]
```diff
@@ -0,0 +1,15 @@
+class CreateCourseEnrolRequests < ActiveRecord::Migration
+  def change
+    create_table :course_enrol_requests do |t|
```
Comment (lowjoel) [inline db/migrate/20150316135303_create_course_enrol_requests.rb]: can we just call it course_registrations or something? enrollment is the list of people in something. you register/apply to enroll into something.

@jsyeo thoughts?

File: db/migrate/20150316135303_create_course_enrol_requests.rb [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+      t.belongs_to :user
+      t.belongs_to :course
```
Comment (lowjoel) [inline db/migrate/20150316135303_create_course_enrol_requests.rb]: order columns from more generic to least generic. course then user.

File: db/migrate/20150316135303_create_course_enrol_requests.rb (hunk 3 above)
Comment (lowjoel) [inline db/migrate/20150316135303_create_course_enrol_requests.rb]: no need deleted_at because there's no soft deletion.

File: spec/factories/course_enrol_requests.rb [hunk 17]
```diff
@@ -0,0 +1,23 @@
+FactoryGirl.define do
+  factory :course_enrol_request, class: 'Course::EnrolRequest' do
```
Comment (lowjoel) [inline spec/factories/course_enrol_requests.rb]: `Course::EnrolRequest.name`

File: spec/models/course/enrol_request_spec.rb (hunk 6 above)
Comment (lowjoel) [inline spec/models/course/enrol_request_spec.rb]: yup.

File: spec/models/course/enrol_request_spec.rb [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+  let!(:instance) { create(:instance) }
+  with_tenant(:instance) do
+    let!(:course) { create(:course) }
+    let!(:user) { create(:user) }
+    let!(:student) { create(:student_enrol_request, course: course) }
+    let!(:teaching_assistant) { create(:teaching_assistant_enrol_request, course: course) }
+    let!(:manager) { create(:manager_enrol_request, course: course) }
+    let!(:owner) { create(:owner_enrol_request, course: course) }
+
+    describe 'creating enrol request of the same user to the same course twice' do
```
Comment (lowjoel) [inline spec/models/course/enrol_request_spec.rb]: context.

File: app/controllers/course/courses_controller.rb (hunk 8 above)
Comment (DIdiHL) [inline app/controllers/course/courses_controller.rb]: Then should the default name be the current user's name?

File: app/models/course.rb (hunk 9 above)
Comment (DIdiHL) [inline app/models/course.rb]: According to their documentation "for belongs_to associations has_many inverse associations are ignored". We can still put inverse_of but it won't be that useful I think.

File: db/migrate/20150316135303_create_course_enrol_requests.rb (hunk 15 above)
Comment (DIdiHL) [inline db/migrate/20150316135303_create_course_enrol_requests.rb]: Though I believe enrol can mean to register someone for a class, I think we can call it registration. I am planning to name the other two ways as mass invitation and secret code invitation. So registration looks to be a good fit as it shows the direction more clearly.

File: app/models/course_user.rb (hunk 13 above)
Comment (DIdiHL) [inline app/models/course_user.rb]: Given that we can get scope like behaviour by .<role> do we still want the scope? Well enum does not support plural form.

**Comment by lowjoel** [comment]
Before continuing, @DIdiHL please fix the complaints from Hound.

File: app/controllers/course/courses_controller.rb [hunk 19]
```diff
@@ -8,7 +8,7 @@ def new #:nodoc:
   end
 
   def create #:nodoc:
-    @course.course_users.build(user: current_user, name: 'name', role: :owner)
+    @course.course_users.build(user: current_user, role: :owner, name: current_user.name)
```
Comment (lowjoel) [inline app/controllers/course/courses_controller.rb]: ...why was this changed?

File: app/models/course.rb [hunk 20]
```diff
@@ -9,6 +9,7 @@ class Course < ActiveRecord::Base
   belongs_to :creator, class_name: User.name
   has_many :course_users, inverse_of: :course, dependent: :destroy
   has_many :users, through: :course_users
+  has_many :registrations
```
Comment (lowjoel) [inline app/models/course.rb]: `dependent: `?

File: app/models/course/registration.rb [hunk 21]
```diff
@@ -0,0 +1,61 @@
+class Course::Registration < ActiveRecord::Base
+  stampable
+
+  scope :staff, -> { where(role: [roles[:teaching_assistant], roles[:manager], roles[:owner]]) }
```
Comment (lowjoel) [inline app/models/course/registration.rb]: you might want to make the `Roles` type in `CourseUser` return you an array of roles that match the given characteristics.

Also, maybe we might want to reason out the categories of users. What do you propose?

File: app/models/course/registration.rb [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+  # @options approve: If specified a true equivalent value, the Registration#approve method will
+  #                   be called. Default to false.
+  #          process_all: if specified a true equivalent value, all registration to the course
+  #                       will be processed. Default to false.
+  #          role: If specified a value in the Registration.roles enum, only registrations of that
+  #                role will be processed. Other IDs are ignored. Default to nil.
+  # @return [Hash] If successful, the hash contains an array of processed IDs with the key -
+  #                'processed_ids'. Otherwise it contains an ActiveModel::Errors with the key -
+  #                'errors'.
+  def self.process(course, registration_ids, options = {})
```
Comment (lowjoel) [inline app/models/course/registration.rb]: Static methods are not good for this purpose.

I also do not think this function is good, abstraction wise. Do not pass IDs around because that breaks abstraction barriers.

File: app/models/course/registration.rb [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+      processed_ids << registration.id
+      registration.destroy!
+    end
+    { processed_ids: processed_ids }
+  end
+
+  def self.registrations_to_be_processed(course, registration_ids, options = {})
+    conditions = {}
+    conditions[:id] = registration_ids unless options[:process_all]
+    options[:role] = %w(teaching_assistant manager owner) if options[:role].to_s == 'staff'
```
Comment (lowjoel) [inline app/models/course/registration.rb]: I prefer constant arrays.

File: app/models/user.rb [hunk 24]
```diff
@@ -10,6 +10,7 @@ class User < ActiveRecord::Base
   has_many :instances, through: :instance_users
   has_many :course_users, inverse_of: :user, dependent: :destroy
   has_many :courses, through: :course_users
+  has_many :registrations
```
Comment (lowjoel) [inline app/models/user.rb]: `dependent: `?

File: db/migrate/20150316140954_create_versions.rb [hunk 25]
```diff
@@ -0,0 +1,13 @@
+class CreateVersions < ActiveRecord::Migration
```
Comment (lowjoel) [inline db/migrate/20150316140954_create_versions.rb]: Split gem initialisation to separate PRs.

File: app/controllers/course/courses_controller.rb (hunk 19 above)
Comment (DIdiHL) [inline app/controllers/course/courses_controller.rb]: Should the default name of a user be 'name'?

File: app/models/course/registration.rb (hunk 21 above)
Comment (DIdiHL) [inline app/models/course/registration.rb]: You mean adding a method that returns an array of role numbers with given role names ? What about `roles.values_at(:teaching_assistant, :manager, :owner)?

I think the roles are just enough. The name of manager may be changed to teacher or lecturer.

File: app/models/course/registration.rb (hunk 23 above)
Comment (DIdiHL) [inline app/models/course/registration.rb]: Hound complains about constant arrays and suggests %w. Maybe the setting is wrong?

File: app/models/course/registration.rb [hunk 26]
```diff
@@ -0,0 +1,29 @@
+class Course::Registration < ActiveRecord::Base
+  stampable
+
+  scope :staff, -> { where(role: roles.values_at(:teaching_assistant, :manager, :owner)) }
```
Comment (lowjoel) [inline app/models/course/registration.rb]: can you try to find a way such that we can expand this staff keyword to this array? Maybe augment the CourseUser::Roles enum. not sure if it's possible.

Comment (DIdiHL) [inline app/models/course/registration.rb]: Did you mean when Course::Registration.staff, it returns something like [1, 2, 3] where the 1, 2 and 3 are role numbers?

Comment (lowjoel) [inline app/models/course/registration.rb]: But in the `CourseUser.roles` hash. I'm not sure if you're allowed to modify the enumeration.

Comment (DIdiHL) [inline app/models/course/registration.rb]: Like `CourseUser.roles.staff`? That's probably not possible.

Comment (lowjoel) [inline app/models/course/registration.rb]: have you tried? CourseUser.roles[:staff] is fine with me too.

Comment (DIdiHL) [inline app/models/course/registration.rb]: Then a way to do it is to add the wanted key, array pair to the roles enum. I can write a module and add a scoped_enum that calls enum and does the addition then let the class extend the module. Is that fine?

Comment (lowjoel) [inline app/models/course/registration.rb]: ...or maybe it's just simpler to declare a fixed hash containing the role categories, maybe `CourseUser.role_categories` so we use `CourseUser.role_categories[:staff]`, if that reduces the likelihood of messing ActiveRecord up.

Comment (DIdiHL) [inline app/models/course/registration.rb]: What about this?
https://gist.github.com/26c7aebbfeb6d7f9328e.git
Well, we can also do the simpler way. The hash returned by `CourseUser.roles` is just a normal hash where we can add anything. So we can even just manually add those extra things to the roles. However I can foresee that methods like `CourseUser.role_scopes` and `course_user.staff?` may be useful.

**Comment by lowjoel** [comment]
I like the direction, but I'm not sure if we're reinventing the wheel here. Maybe this should be implemented externally by a gem? The existing gems don't seem to do this though.

Also  I think I need to split the extensions libraries... it's getting incredibly cluttered.

**Comment by DIdiHL** [comment]
I did some preliminary search with no avail. Probably there is no existing gem. I can package it as a gem though. If left this state, it generates a lot of rubocop offenses.

File: app/models/course/registration.rb (hunk 26 above)
Comment (allenwq) [inline app/models/course/registration.rb]: Find a gem which might be useful: https://github.com/brainspec/enumerize

it supports scopes like:

```
   User.with_role(:teaching_assistant, :manger, :owner)
```

And also helps with I18n. It does not support the categories things, but I think it's fine? just define a new scope with the code above.

File: app/models/course/registration.rb [hunk 27]
```diff
@@ -0,0 +1,35 @@
+class Course::Registration < ActiveRecord::Base
```
Comment (lowjoel) [inline app/models/course/registration.rb]: I'm starting to wonder if this is necessary.

Isn't this just a Course::CourseUser that is not enabled? This should be a state machine.

File: app/views/enrollment_mailer/new_registration.html.slim [hunk 28]
```diff
@@ -0,0 +1,7 @@
+p = t('.greeting_format') % { name: @lecturer.name }
```
Comment (lowjoel) [inline app/views/enrollment_mailer/new_registration.html.slim]: isn't this supposed to be a parameter for `t()`?

File: app/models/course/registration.rb (hunk 27 above)
Comment (DIdiHL) [inline app/models/course/registration.rb]: Sounds legit.

File: app/models/course/registration.rb (hunk 26 above)
Comment (DIdiHL) [inline app/models/course/registration.rb]: Not sure if it the thing Joel wants. Probably he wants a name for the group of enum values? Let's discuss in the next meeting.

**Comment by allenwq** [comment]
@DIdiHL Is this PR ready for review ? and  how's the scoped enum gem ?

**Comment by DIdiHL** [comment]
Yes, it's ready. The gem is not ready yet. I think it'll be added in another PR.

File: app/models/course_user.rb [hunk 29]
```diff
@@ -1,5 +1,32 @@
...
+  end
+
+  def staff_of_the_course
+    course.course_users.staff
+  end
+
+  # Approve the registration record by creating a CourseUser of the user to the course with the
+  # requested role.
+  def approve
+    update(status: CourseUser.statuses[:approved])
```
Comment (allenwq) [inline app/models/course_user.rb]: I don't understand why need to reference <code> CourseUser.statuses </code> to get the status

**Comment by DIdiHL** [comment]
It needs to be the integer value. The alternative will be hard coding it but I don't think it looks good.

File: app/models/course_user.rb (hunk 29 above)
Comment (allenwq) [inline app/models/course_user.rb]: <code>update(status: :approved)</code> does not work ?

Comment (DIdiHL) [inline app/models/course_user.rb]: Nope. The column is integer type, and update is not aware of the enum.

Comment (allenwq) [inline app/models/course_user.rb]: I think rails should have helper for this. have you tried ?

Comment (DIdiHL) [inline app/models/course_user.rb]: Yup, initially I passed in a Symbol and it complained. That's why I changed to refer to the end.

Comment (allenwq) [inline app/models/course_user.rb]: <code> @course.course_users.build(user: current_user, name: 'name', role: :owner, status: :approved) </code> but why here it works ?

Comment (DIdiHL) [inline app/models/course_user.rb]: Hmm, lemme check when I'm back at a computer.

Comment (DIdiHL) [inline app/models/course_user.rb]: Oh it works! Will change to that.

Comment (allenwq) [inline app/models/course_user.rb]: :+1:

File: app/mailers/enrollment_mailer.rb [hunk 30]
```diff
@@ -0,0 +1,21 @@
+class EnrollmentMailer < ActionMailer::Base
+  helper_method :registration_notice
+  default from: 'noreply@coursemology.org'
```
Comment (lowjoel) [inline app/mailers/enrollment_mailer.rb]: this should be configurable.

File: app/mailers/enrollment_mailer.rb [hunk 31]
```diff
@@ -0,0 +1,21 @@
...
+  def new_registration(registration, lecturer)
+    @course = registration.course
+    @role = registration.role
+    @user = registration.user
+    @lecturer = lecturer
+    mail(to: lecturer.email, subject: t('.subject'))
+  end
+
+  def registration_notice
+    t('.notice_format') % {
```
Comment (lowjoel) [inline app/mailers/enrollment_mailer.rb]: I think those are supposed to be parameters to `t()`

File: app/models/course_user.rb [hunk 32]
```diff
@@ -1,5 +1,32 @@
...
+  enum role: { student: 0, teaching_assistant: 1, manager: 2, owner: 3 },
+       status: { pending: 0, approved: 1 }
+
+  scope :staff, -> { where(role: [roles[:teaching_assistant], roles[:manager], roles[:owner]]) }
+
+  def staff?
+    teaching_assistant? || manager? || owner?
+  end
+
+  def staff_of_the_course
```
Comment (lowjoel) [inline app/models/course_user.rb]: ...I don't think this method belongs here.

File: app/models/course_user.rb [hunk 33]
```diff
@@ -1,5 +1,32 @@
 class CourseUser < ActiveRecord::Base
+  stampable
   belongs_to :user, inverse_of: :course_users
   belongs_to :course, inverse_of: :course_users
-  enum role: { student: 0, teaching_assistant: 1, manager: 2, owner: 3 }
+
+  enum role: { student: 0, teaching_assistant: 1, manager: 2, owner: 3 },
+       status: { pending: 0, approved: 1 }
```
Comment (lowjoel) [inline app/models/course_user.rb]: This ought to be a workflow. see geekq/workflow@403a9e44bf49f4d154156d5701e3d67b115ed6da

File: spec/mailers/enrollment_mailer_spec.rb [hunk 34]
```diff
@@ -0,0 +1,40 @@
...
+    ActsAsTenant.current_tenant = Instance.default
+  end
+
+  let!(:course) { create(:course) }
+  let!(:student) { create(:user) }
+  let!(:lecturer) { create(:course_owner, course: course, user: course.creator, status: :approved) }
+  let!(:registration) { create(:course_student, course: course, user: student) }
+
+  def translate(mail_action, mail_field)
+    I18n.t("enrollment_mailer.#{mail_action}.#{mail_field}", raise: true)
```
Comment (lowjoel) [inline spec/mailers/enrollment_mailer_spec.rb]: This doesn't seem right either... I don't think it's meant to be used like this :P

File: app/models/course_user.rb [hunk 35]
```diff
@@ -1,5 +1,32 @@
...
+
+  scope :staff, -> { where(role: [roles[:teaching_assistant], roles[:manager], roles[:owner]]) }
```
Comment (lowjoel) [inline app/models/course_user.rb]: define a private `staff_roles`, then use `where(staff_roles)`.

File: app/models/course_user.rb [hunk 36]
```diff
@@ -1,5 +1,32 @@
...
+
+  def staff?
+    teaching_assistant? || manager? || owner?
```
Comment (lowjoel) [inline app/models/course_user.rb]: `self.class.staff_roles.contains(role)`

**Comment by allenwq** [comment]
@lowjoel  Have a look thanks ! I'll config mailer in anther PR.

File: Gemfile [hunk 37]
```diff
@@ -138,5 +138,8 @@ gem 'carrierwave', github: 'carrierwaveuploader/carrierwave'
 # Required by carrierwave, for image resizing
 gem 'mini_magick'
 
+# A Ruby state machine library
+gem 'workflow'
```
Comment (lowjoel) [inline Gemfile]: Move to place together with the model gems.

File: app/controllers/course/courses_controller.rb [hunk 38]
```diff
@@ -6,9 +6,10 @@ def new #:nodoc:
   end
 
   def create #:nodoc:
-    @course.course_users.build(user: current_user, name: 'name', role: :owner)
+    creator = @course.course_users.build(user: current_user, name: 'name', role: :owner)
 
     if @course.save
+      creator.approve!
```
Comment (lowjoel) [inline app/controllers/course/courses_controller.rb]: I believe workflow lets you set the state of the object when creating the object.

File: app/models/course_user.rb [hunk 39]
```diff
@@ -1,5 +1,33 @@
 class CourseUser < ActiveRecord::Base
+  include Workflow
+  workflow do
+    state :pending do
+      event :approve, transitions_to: :approved
+      event :reject, transitions_to: :rejected
+    end
+    state :approved
+    state :rejected
```
Comment (lowjoel) [inline app/models/course_user.rb]: there should be a handler to delete the record right?

File: app/models/course_user.rb [hunk 40]
```diff
@@ -1,5 +1,33 @@
...
+
   enum role: { student: 0, teaching_assistant: 1, manager: 2, owner: 3 }
+
+  scope :staff, -> { where(role: staff_roles) }
+
+  # Returns staff roles
+  #
+  # @return [Array] array of staff role enums
+  def self.staff_roles
+    [roles[:teaching_assistant], roles[:manager], roles[:owner]]
```
Comment (lowjoel) [inline app/models/course_user.rb]: This should be a constant.

Comment (lowjoel) [inline app/models/course_user.rb]: In the singleton object.

File: app/models/course_user.rb [hunk 41]
```diff
@@ -1,5 +1,33 @@
...
+  end
+
+  # Test whether this course_user is a staff (i.e. teaching_assistant, manager or owner)
+  #
+  # @return [Boolean] True if course_user is a staff
+  def staff?
+    self.class.staff_roles.include?(self[:role])
```
Comment (lowjoel) [inline app/models/course_user.rb]: `CourseUser.STAFF_ROLES...`

How come this is called CourseUser and not Course::User :confused:

File: db/migrate/20150512015621_add_userstamps_to_course_users.rb [hunk 42]
```diff
@@ -0,0 +1,14 @@
+class AddUserstampsToCourseUsers < ActiveRecord::Migration
```
Comment (lowjoel) [inline db/migrate/20150512015621_add_userstamps_to_course_users.rb]: I'm not sure if this makes sense. The courseuser creator will be always.... null

File: app/controllers/course/courses_controller.rb (hunk 38 above)
Comment (allenwq) [inline app/controllers/course/courses_controller.rb]: I didn't see their official API about this, but we can directly do <code>build(workflow_state: 'approved')</code>

File: app/models/course_user.rb (hunk 39 above)
Comment (allenwq) [inline app/models/course_user.rb]: Yes, thanks !

File: app/models/course_user.rb (hunk 40 above)
Comment (allenwq) [inline app/models/course_user.rb]: Sorry, but why need to declare it in the singleton object ?

File: app/models/course_user.rb (hunk 41 above)
Comment (allenwq) [inline app/models/course_user.rb]: OK, I agree, Actually we can rename it to Course::User without touching the db table.

File: db/migrate/20150512015621_add_userstamps_to_course_users.rb (hunk 42 above)
Comment (allenwq) [inline db/migrate/20150512015621_add_userstamps_to_course_users.rb]: Oh yeah, seems that the creator will always be the one who submit the enrol request ?

Comment (lowjoel) [inline db/migrate/20150512015621_add_userstamps_to_course_users.rb]: yes
